cmd/ingress/main.go

```diff
@@ -19,62 +19,132 @@ package main
+	if err := envconfig.Process("", &env); err != nil {
+		log.Fatal("Failed to process env var", zap.Error(err))
+	}
+
+	conn, err := amqp.Dial(env.BrokerURL)
 	if err != nil {
 		log.Fatalf("failed to connect to RabbitMQ: %s", err)
 	}
 	defer conn.Close()
 
-	channel, err := conn.Channel()
+	env.channel, err = conn.Channel()
 	if err != nil {
 		log.Fatalf("failed to open a channel: %s", err)
 	}
-	defer channel.Close()
+	defer env.channel.Close()
+
+	env.logger = logging.FromContext(context.Background())
+
+	connectionArgs := kncloudevents.ConnectionArgs{
+		MaxIdleConns:        defaultMaxIdleConnections,
+		MaxIdleConnsPerHost: defaultMaxIdleConnectionsPerHost,
+	}
+	kncloudevents.ConfigureConnectionArgs(&connectionArgs)
 
-	c, err := cloudevents.NewDefaultClient()
+	receiver := kncloudevents.NewHTTPMessageReceiver(env.Port)
+
+	if err := receiver.StartListen(context.Background(), &env); err != nil {
+		log.Fatalf("failed to start listen, %v", err)
+	}
+}
+
+func (env *envConfig) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	// validate request method
+	if request.Method != http.MethodPost {
+		env.logger.Warn("unexpected request method", zap.String("method", request.Method))
+		writer.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	// validate request URI
+	if request.RequestURI != "/" {
+		writer.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	ctx := request.Context()
+
+	message := cehttp.NewMessageFromHttpRequest(request)
+	defer message.Finish(nil)
+
+	event, err := binding.ToEvent(ctx, message)
+	if err != nil {
+		env.logger.Warn("failed to extract event from request", zap.Error(err))
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// run validation for the extracted event
+	validationErr := event.Validate()
+	if validationErr != nil {
+		env.logger.Warn("failed to validate extracted event", zap.Error(validationErr))
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	statusCode, err := env.send(event)
 	if err != nil {
-		log.Fatalf("failed to create cloudevents client, %v", err)
+		env.logger.Error("failed to send event,", err)
 	}
+	writer.WriteHeader(statusCode)
+}
 
-	log.Fatal(c.StartReceiver(context.Background(), func(event cloudevents.Event) {
-		log.Printf("received: %s", event)
-		bytes, err := json.Marshal(event)
-		if err != nil {
-			log.Fatalf("failed to marshal event, %v", err)
-		}
-		headers := amqp.Table{
-			"type":    event.Type(),
-			"source":  event.Source(),
-			"subject": event.Subject(),
-		}
-		for key, val := range event.Extensions() {
-			headers[key] = val
-		}
-		err = channel.Publish(
-			exchangeName,
-			"",    // routing key
-			false, // mandatory
-			false, // immediate
-			amqp.Publishing{
-				Headers:     headers,
-				ContentType: "application/json",
-				Body:        bytes,
-			})
-		if err != nil {
-			log.Fatal("failed to publish message")
-		}
-	}))
+func (env *envConfig) send(event *cloudevents.Event) (int, error) {
```

Comment:
I think I'd prefer the signature to have context, channel and exchangeName in it rather than pull them from the envConfig (see above comment).

Reply:
That way at least the context from the request could be wired through that you get on L97?

Reply:
I was going to wire in the context, but I did not see a context being used here so I skipped it

Reply:
I did it this way because I already have envconf overloaded with the `func (env *envConfig) ServeHTTP` function